language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -276,136 +357,83 @@ computeCek con t@(Var _ varName)   = do
-returnCek (FrameUnwrap : con) dat = case dat of
-    IWrap _ _ _ term -> returnCek con term
-    term             ->
-        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just term
+    => Context uni -> CekVal uni -> CekM uni (Plain Term uni)
+--- Instantiate all the free variable of the resulting term in case there are any.
+-- . ◅ V           ↦  [] V
+returnCek [] val = pure $ void $ dischargeCekVal val
+-- s , {_ A} ◅ abs α M  ↦  s ; ρ ▻ M [ α / A ]*
+returnCek (FrameTyInstArg ty : ctx) fun = instantiateEvaluate ctx ty fun
+-- s , [_ (M,ρ)] ◅ V  ↦  s , [V _] ; ρ ▻ M
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
+    -- funVarEnv <- getEnv
+    withEnv argVarEnv $ computeCek (FrameApplyFun fun : ctx) arg
+-- s , [(lam x (M,ρ)) _] ◅ V  ↦  s ; ρ [ x  ↦  V ] ▻ M
+returnCek (FrameApplyFun fun : ctx) arg = do
+    applyEvaluate ctx fun arg
+-- s , wrap A B _ ◅ V  ↦  s ◅ wrap A B V
+returnCek (FrameIWrap ex pat arg : ctx) val =
+    returnCek ctx $ VIWrap ex pat arg val
+-- s , unwrap _ ◅ wrap A B V  ↦  s ◅ V
+returnCek (FrameUnwrap : ctx) val =
+    case val of
+      VIWrap _ _ _ v -> returnCek ctx v
+      _              ->
+        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just val
 
 -- | Instantiate a term with a type and proceed.
 -- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
--- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
+-- iterated application of a 'BuiltinName' to a list of 'Value's and if it is,
 -- apply the term to the type via 'TyInst'.
 instantiateEvaluate
     :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> Type TyName uni ExMemory -> WithMemory Term uni -> CekM uni (Plain Term uni)
-instantiateEvaluate con _ (TyAbs _ _ _ body) = computeCek con body
-instantiateEvaluate con ty fun
-    | isJust $ termAsPrimIterApp fun = returnCek con $ TyInst (memoryUsage () <> memoryUsage fun <> memoryUsage ty) fun ty
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just fun
-
-{- Note [Saved mapping example]
-Consider a polymorphic built-in function @id@, whose type signature on the Plutus side is
-
-    id : all a. a -> a
-
-Notation:
-
-- the variable environment is denoted as @{ <var> :-> (<env>, <value>), ... }@
-- the context is denoted as a Haskell list of 'Frame's
-- 'computeCek' is denoted as @(|>)@
-- 'returnCek' is denoted as @(<|)@
-
-When evaluating
+    => Context uni -> Type TyName uni ExMemory -> CekVal uni -> CekM uni (Plain Term uni)
+instantiateEvaluate ctx _ (VTyAbs _ _ _ body env) = withEnv env $ computeCek ctx body -- FIXME: env?
+instantiateEvaluate ctx ty (VBuiltin ex argEnv bn count tyargs args) =
+    case args of
+      [] -> returnCek ctx $ VBuiltin ex argEnv bn count (ty:tyargs) args
+      _  -> error "Builtin instantiation after term argument"
```

Comment:
Let's create a `CekInternalError` and add this error there together with `UnknownDynamicBuiltinNameError`.

But do we need it here? Can we just allow interleaving type and term arguments?

Reply:
I've now renamed `countArgs` to `countTypeAndTermArgs` and modified it to do what the name suggests.  Here we just decrement the count and add the type to tyargs in all cases.  We still run the risk of having type and term arguments wrongly interleaved, but the typechecker should detect that.  I'm tempted to store the type scheme in `VBuiltin` and strip off the head when we're applying or instantiating, making sure that we've got the right kind of argument.  That would reduce efficiency, but maybe that doesn't matter so much for the typed machine.

Reply:
Damn.  Counting both types of arguments breaks the CK machine because when it got a builtin application it was assuming that it was fully instantiated and it just checked the number of term arguments using `termAsPrimIterApp`. I'll have to do something better than that.  I think I'm going to need a special frame for builtins so that it can remember the number of type and term arguments it's seen so far.